torchao/sparsity/dynamic_quant_sparse.py

```diff
@@ -0,0 +1,300 @@
+    out_dtype,
+):
+
+    assert (
+        x_vals_int8.dtype == torch.int8
+    ), f"x dtype {x_vals_int8.dtype} not yet supported"
+    assert (
+        w_vals_int8.dtype == torch.int8
+    ), f"w dtype {w_vals_int8.dtype} not yet supported"
+    # assert w_scales.dtype == out_dtype, f'{w_scales.dtype} does not match {out_dtype}'
+
+    tmp = x_vals_int8.reshape(-1, x_vals_int8.shape[-1]).contiguous()
+
+    assert x_scales.dtype in [
+        torch.float,
+        torch.bfloat16,
+    ], f"x_scales needs to be a torch.float32 or torch.bfloat16 but got {x_scales.dtype}"
+
+    y_dot_bf16_w_scales_fused = torch._cslt_sparse_mm(
+        w_vals_int8, tmp.t(), out_dtype=torch.bfloat16
+    ).t()
+    y = (y_dot_bf16_w_scales_fused * x_scales.reshape(-1, 1) * w_scales).reshape(
+        *x_vals_int8.shape[:-1], y_dot_bf16_w_scales_fused.shape[-1]
+    )
+    y = y.to(out_dtype)
+
+    return y
+
+
+def sparse_quant_int8_cutlass_matmul(
+    x_vals_int8,
+    x_scales,
+    w_vals_int8,
+    w_meta_int32,
+    w_scales,
+    out_dtype,
+):
+    assert (
+        x_vals_int8.dtype == torch.int8
+    ), f"x dtype {x_vals_int8.dtype} not yet supported"
+    assert (
+        w_vals_int8.dtype == torch.int8
+    ), f"w dtype {w_vals_int8.dtype} not yet supported"
+    assert w_scales.dtype == out_dtype, f"{w_scales.dtype} does not match {out_dtype}"
+    assert w_meta_int32.dtype == torch.int32, f"{w_meta_int32.dtype} not yet supported"
+
+    tmp = x_vals_int8.reshape(-1, x_vals_int8.shape[-1]).contiguous()
+
+    assert x_scales.dtype in [
+        torch.float,
+        torch.bfloat16,
+    ], f"x_scales needs to be a torch.float32 or torch.bfloat16 but got {x_scales.dtype}"
+
+    y_dot_int32 = torch._sparse_semi_structured_linear(
+        tmp, w_vals_int8, w_meta_int32.view(torch.int32), out_dtype=torch.int32
+    )
+    y = (y_dot_int32 * x_scales.reshape(-1, 1) * w_scales).reshape(
+        *x_vals_int8.shape[:-1], y_dot_int32.shape[-1]
+    )
+    y = y.to(out_dtype)
+    return y
+
+class Int8DynamicallyQuantized24CusparseltLinearFuseMulWeight(
+    Int8DynamicallyQuantizedLinearWeight
+):
+
+    @staticmethod
+    def _quantized_op(act_mat, w_qtensor, bias):
+        return sparse_quant_int8_dynamic_linear(
+            act_mat, w_qtensor.int_data, None, w_qtensor.q_scales, bias, act_mat.dtype,
+            fuse_mul=True
+        )
+
+    @classmethod
+    def from_float(cls, input_float, qmin=-8, qmax=7):
+
+        assert input_float.is_cuda
+
+        w_int_repr, w_scales, _ = dynamically_quantize_per_channel(
+            input_float, qmin, qmax, torch.int8
+        )
+
+        int_data = w_int_repr.contiguous()
+        int_data = torch._cslt_compress(int_data)
+
+        return cls(
+            int_data,
+            w_scales,
+            False,
+            input_float.shape,
+            dtype=input_float.dtype,
+        )
+
+
+class Int8DynamicallyQuantizedSemiStructuredSparseLinearWeight(QuantizedLinearWeightBase):
```

Comment:
is sparsity also implemented with tensor subclass? I thought we should be able to compose them in some way?

Reply:
We don't have nested subclassing support currently for tracing, so we can't compose them currently :( hence why we're landing in prototype.

I can tag you in the issue i'll make to raise this for core.